lib/rocksdb.rb

```diff
@@ -1,61 +1,72 @@
 require "rocksdb/RocksDB" # the c extension
 require "rocksdb/ruby/version"
 
+require 'forwardable'
+
 module RocksDB
-  class DBError < StandardError; end
-  class DB
+  class Error < StandardError; end
+  class ReadOnly < RocksDB::Error; end
+  class StatusError < RocksDB::Error; end
+  class DatabaseClosed < RocksDB::Error; end
+  class IteratorClosed < RocksDB::Error; end
+
+  class Iterator
     include Enumerable
+  end
 
-    @@cache = {}
-    
-    class << self
-      def get_instance *args
-        readonly = !!(args[1] && args[1][:readonly])
-        key = args[0]
-        
-        if readonly
-          return new(*args)
-        end
-        unless @@cache[key]
-          @@cache[key] = new(*args)
-        end
-        @@cache[key]
-      end
+  class << self
+    def open(db_path, db_options = "")
+      ::RocksDB::DB.new(db_path, db_options, readonly: false)
     end
-    
-    def initialize *args
-      readonly = !!(args[1] && args[1][:readonly])
-      @key = args[0]
-      if !readonly and @@cache[@key]
-        __initialize2(*args)
-        raise DBError.new("error #{@key.to_s} alread open")
-      end
-      __initialize(*args)
-      unless readonly
-        @@cache[@key] = self
-      end
+
+    def open_readonly(db_path, db_options = "")
+      ::RocksDB::DB.new(db_path, db_options, readonly: true)
     end
+  end
+
+  class DB
+    extend Forwardable
 
-    def close
-      @@cache.delete(@key)
-      __close
+    def initialize(path, rocksdb_options = "", options = {})
+      is_readonly = options[:readonly] || false
+
+      if rocksdb_options.is_a? Hash
+        is_readonly = rocksdb_options.delete(:readonly)
+
+        rocksdb_options = rocksdb_options.map do |key, value|
+          [key, value].join("=")
+        end.join(";")
+      end
+
+      __initialize(path, is_readonly, rocksdb_options.to_s)
     end
-    
-    alias :includes? :exists?
-    alias :contains? :exists?
-    alias :member? :exists?
-    alias :[] :get
-    alias :[]= :put
-    alias :close! :close
-
-    def each(&block)
-      if block_given?
-        self.each_with_index do |key, value|
-          block.call(value)
-        end
+
+    def get(*args)
+      args.flatten!
+
+      if args.size == 1
+        get_one args.first
       else
-        self.iterator
+        get_many args
       end
     end
+
+    alias_method :includes?, :exists?
+    alias_method :contains?, :exists?
+    alias_method :member?, :exists?
+    alias_method :member?, :exists?
+    alias_method :[], :get
+    alias_method :[]=, :put
```

Comment:
I'm not 100% sure about this. I understand it's Ruby-ish and cool, but in general I'm not a fan of having multiple ways to do exactly the same thing.

On related note, `database['something']` is not a clear concept at the first glance - is it fetching data or maybe some configuration option? `database.get` is rather clear in that regard.

Reply:
I kept this part of interface as it was: 

https://github.com/isamu/rocksdb-ruby/blob/abe4be82d6d2291f10a30c426af66a9f8fa55326/lib/rocksdb.rb#L47-L48

It's a quite common thing in the ruby world to have multiple names for the same thing to make code more readable. 

I understand your concern about `[]`, and there is no consistency between other gems. Few popular ones ([moneta-rb](https://github.com/moneta-rb/moneta), [daybreak](https://propublica.github.io/daybreak/), [riak](https://github.com/basho/riak-ruby-client))  have `[]` and `[]=` as well as `get` and `set` (I think it worth adding `set` along with `put`)

I don't have strong opinion about that, and will deprecate them if needed.

Reply:
> It's a quite common thing in the ruby world to have multiple names for the same thing to make code more readable.

Yes, I'm aware of that, but I still think that smaller API is a better API 😉 Then, similar to you, it's not a very strong opinion. Perhaps we need more people to express their opinion.